### PR TITLE
Add StreamConsumer wrapper and methods to response

### LINKED ISSUE #21476: Support asynchronous WebAssembly compilation
This is a web assembly game demo.

STR:
- Open https://www.funkykarts.rocks/demo.html 

Expected:
- you have fun!

Observed:
- You are sad: `WebAssembly Promise APIs not supported in this runtime`

I guess these are the streaming apis from https://webassembly.github.io/spec/web-api/index.html

### PR DESCRIPTION
**Pull request by ridhimrastogi**
Add Streamconsumer wrapper to Response

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #21476

### CODE REVIEW

**Comment by ridhimrastogi** [comment]
The commit fails as the newly added functions, set_stream_consumer and stream_chunk aren't being called yet.

File: components/script/dom/response.rs [hunk 1]
```diff
@@ -441,11 +445,24 @@ impl Response {
         }
     }
 
+    pub fn set_stream_consumer(&self, sc: Option<StreamConsumer>) {
+        *self.stream_consumer.borrow_mut() = sc;
+    }
+
+    pub fn stream_chunk(&self, stream: &[u8]) {
+        if let Some(stream_consumer) = self.stream_consumer.borrow_mut().take() {
```
Comment (jdm) [inline components/script/dom/response.rs]: take() consumes the value, so we want  as_ref() instead.

File: components/net/tests/fetch.rs [hunk 2]
```diff
@@ -138,6 +138,7 @@ fn test_fetch_blob() {
         fn process_response(&mut self, _: &Response) {}
         fn process_response_chunk(&mut self, chunk: Vec<u8>) {
             self.buffer.extend_from_slice(chunk.as_slice());
+            self.stream_chunk(buffer.as_slice());
```
Comment (jdm) [inline components/net/tests/fetch.rs]: This is a unit test for unrelated code; we should be modifying components/script/fetch.rs instead.

**Review by jdm — CHANGES_REQUESTED** [review]
Good start!

File: components/script/dom/response.rs (hunk 1 above)
Comment (ridhimrastogi) [reply components/script/dom/response.rs]: Done

File: components/net/tests/fetch.rs (hunk 2 above)
Comment (ridhimrastogi) [reply components/net/tests/fetch.rs]: Oops, had both files open and made changes in the wrong one. Fixed it now

File: components/script/fetch.rs [hunk 3]
```diff
@@ -261,6 +261,7 @@ impl FetchResponseListener for FetchContext {
 
     fn process_response_chunk(&mut self, mut chunk: Vec<u8>) {
         self.body.append(&mut chunk);
+        self.stream_chunk(chunk.as_slice());
```
Comment (jdm) [inline components/script/fetch.rs]: Since this is a method on FetchContext and not Response, you need to use `aelf.respinse_object.root()` to obtain a `DomRoom<Response>` object that you can use to call Response methods.

**Comment by jdm** [comment]
I think the only missing piece now is calling Response::set_stream_consumer from the stream consumer callback?

File: components/script/fetch.rs (hunk 3 above)
Comment (ridhimrastogi) [reply components/script/fetch.rs]: Done

**Comment by ridhimrastogi** [comment]
> I think the only missing piece now is calling Response::set_stream_consumer from the stream consumer callback?

Done

**Comment by ridhimrastogi** [comment]
@jdm I see some changes in the test results when I run  
`./mach test-wpt tests/wpt/web-platform-tests/wasm/webapi/`.
Do I need to update test expectations for these changes?

**Comment by jdm** [comment]
Yep!

File: components/script/fetch.rs [hunk 4]
```diff
@@ -261,6 +261,7 @@ impl FetchResponseListener for FetchContext {
 
     fn process_response_chunk(&mut self, mut chunk: Vec<u8>) {
         self.body.append(&mut chunk);
+        self.response_object.root().stream_chunk(chunk.as_slice());
```
Comment (jdm) [inline components/script/fetch.rs]: It turns out that `self.body.append(&mut chunk);` empties the passed vector, which is why a bunch of tests that I would expect to pass are failing with these changes. If we reverse the order of these lines, then the `compileStreaming` tests from /wasm/webapi/contenttype.any.html (and likely other test files as well) should start passing.

**Review by jdm — CHANGES_REQUESTED** [review]
Looks great! With one final change this should be ready to merge :)

File: components/script/fetch.rs (hunk 4 above)
Comment (ridhimrastogi) [reply components/script/fetch.rs]: I have made the changes. Now the mimeType related tests are passing. However, some of the response status tests in ```/tests/wpt/metadata/wasm/webapi/status.any.js.ini``` are failing, which I believe should have passed. Is this expected?

**Comment by jdm** [comment]
I'm fine with merging the changes as-is.
@bors-servo r+

**Comment by CYBAI** [comment]
on both linux and macOS builds:

```
  ▶ ERROR [expected TIMEOUT] /wasm/serialization/module/identity-not-preserved.html
  └   → Unhandled rejection: The object can not be cloned.
```

**Comment by jdm** [comment]
That's #24726; I'll retry when the other pending tasks finish running.

**Comment by jdm** [comment]
@bors-servo retry
* #24726
